googledataprocauthenticator/google.py

```diff
@@ -0,0 +1,319 @@
+        new_exc = BadUserConfigurationException(
+            "Gcloud is not installed. Install the Google Cloud SDK." 
+        )
+        raise new_exc
+    except (subprocess.CalledProcessError, IOError) as caught_exc:
+        new_exc = BadUserConfigurationException(
+            "Failed to obtain access token. Run `gcloud auth login` in your command line "\
+            "to authorize gcloud to access the Cloud Platform with Google user credentials "\
+            "to authenticate. Run `gcloud auth application-default login` to acquire new "\
+            "user credentials to use for Application Default Credentials.")
+        raise new_exc
+
+
+def get_cluster_pool(project_id, region, client):
+    #filter format: status.state = ACTIVE AND clusterName = mycluster AND labels.env = staging AND labels.starred = \*
+    cluster_pool = set()
+    for cluster in client.list_clusters(project_id, region, 'status.state = ACTIVE'):
+        #check component gateway is enabled
+        if (len(cluster.config.endpoint_config.http_ports.values()) != 0):
+            action_list = list()
+            for action in cluster.config.initialization_actions:
+                #check if init action is the livy action with a region that matches the regex pattern [a-z0-9-]+
+                is_livy_action = re.search("gs://goog-dataproc-initialization-actions-[a-z0-9-]+/livy/livy.sh", bucket) is not None                    
+                if is_livy_action:
+                    action_list.append(action.executable_file)
+                    cluster_pool.add(cluster.cluster_name)
+    return cluster_pool
+
+
+def get_component_gateway_url(project_id, region, cluster_name, credentials):
+    """Gets the component gateway url for a cluster name, project id, and region
+
+    Args:
+        cluster_name (str): The cluster name to use for the url
+        project_id (str): The project id to use for the url
+        region (str): The project id to use for the url
+        credentials (google.oauth2.credentials.Credentials): The authorization credentials to
+        attach to requests.
+
+    Returns:
+        str: the component gateway url
+
+    Raises:
+        google.api_core.exceptions.GoogleAPICallError: If the request failed for any reason.
+        google.api_core.exceptions.RetryError: If the request failed due to a retryable error and retry attempts failed.
+        ValueError: If the parameters are invalid.
+    """
+    client = dataproc_v1beta2.ClusterControllerClient(credentials=credentials,
+                       client_options={
+                            'api_endpoint': '{}-dataproc.googleapis.com:443'.format(region)
+                        }
+                    )
+    try:
+        print(get_cluster_pool(project_id, region, client))
+        # if they didn't enter cluster name then we get a cluster from cluster pool 
+        cluster_pool = None
+        if cluster_name is '':
+            #pop random cluster from cluster pool 
```

Comment:
Nit: If you want a random value, consider using `random.choice(cluster_pool)`. Even though sets are unordered, I suspect that `.pop()` will end up returning the same value every time: https://stackoverflow.com/questions/21017188/set-pop-isnt-random

Reply:
Is there a reason why we wouldn't want to get the same cluster everytime? I decided to use pop because I thought it didn't matter which cluster we would get, so is there a benefit to making it explicitly random?

Reply:
I'm mostly afraid that several users will try to connect to the same pool of clusters, and all get assigned to the same cluster. Ideally, they'd be spread out evenly among the clusters